src/libopensc/pkcs15-pubkey.c

```diff
@@ -311,6 +311,10 @@ int sc_pkcs15_decode_pukdf_entry(struct sc_pkcs15_card *p15card,
 	memset(gostr3410_params, 0, sizeof(gostr3410_params));
 
 	r = sc_asn1_decode(ctx, asn1_pubkey, *buf, *buflen, buf, buflen);
+	if (r < 0) {
+		free(der->value);
```

Comment:
Has `der->value` reallly been reported as leaking? It's effectively [`obj->content`](https://github.com/OpenSC/OpenSC/blob/45d7338351b4e4e23b66a89c4801d959118066aa/src/libopensc/pkcs15-pubkey.c#L220), which should be freed [here](https://github.com/OpenSC/OpenSC/blob/8fea658fe9b5bdf51de7e321a1a549056ba4b150/src/tests/fuzzing/fuzz_pkcs15_decode.c#L61)...

Reply:
The reports are too vague and reproducing it in docker does not always yield the expected results.

It is freed only if the `obj->type` is correctly set, which is not the case if we fail here. Type is set just few lines below this.

Reply:
No, it also gets [`free`d when having an arbitrary type](https://github.com/OpenSC/OpenSC/blob/8fea658fe9b5bdf51de7e321a1a549056ba4b150/src/libopensc/pkcs15.c#L2536).

Reply:
The backtrace points directly to this place, and the `der->value` is the only ASN1 member of `SC_ASN1_OCTET_STRING` type with `SC_ASN1_ALLOC` flags. I can reproduce it locally (in oss-fuzz container) to get more verbose backtrace:
```
#0  __GI___libc_malloc (bytes=0) at malloc.c:2902
#1  0x00000000006451a3 in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd79e8, obj=0x224273d "\244!\363", objlen=0, depth=5) at asn1.c:1559
#2  0x0000000000642fae in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd79c0, in=0x224273b "\240", len=5, newp=0x7ffeb7bd6000, len_left=0x7ffeb7bd5fe8, choice=1, depth=5) at asn1.c:1727
#3  0x0000000000642caa in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd7940, in=0x224273b "\240", len=5, newp=0x0, len_left=0x0, choice=0, depth=4) at asn1.c:1698
#4  0x0000000000644e4b in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd78f0, obj=0x224273b "\240", objlen=5, depth=3) at asn1.c:1490
#5  0x0000000000642fae in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd78f0, in=0x2242739 "0\005\240", len=10, newp=0x0, len_left=0x0, choice=0, depth=3) at asn1.c:1727
#6  0x0000000000644e4b in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd6d98, obj=0x2242739 "0\005\240", objlen=10, depth=2) at asn1.c:1490
#7  0x0000000000642fae in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd6d20, in=0x2242722 "0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", len=33, newp=0x0, len_left=0x0, choice=0, depth=2) at asn1.c:1727
#8  0x0000000000645f97 in asn1_decode_p15_object (ctx=0x2242830, in=0x2242722 "0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", len=33, obj=0x7ffeb7bd75d0, depth=1) at asn1.c:1411
#9  0x0000000000645568 in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd76d8, obj=0x2242722 "0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", objlen=33, depth=1) at asn1.c:1632
#10 0x0000000000642fae in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd7660, in=0x2242720 "\240!0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", len=262, newp=0x7ffeb7bd7410, len_left=0x7ffeb7bd73f8, choice=1, depth=1)
    at asn1.c:1727
#11 0x0000000000642caa in asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd7610, in=0x2242720 "\240!0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", len=262, newp=0x7ffeb7bd7d98, len_left=0x7ffeb7bd7d90, choice=0, depth=0)
    at asn1.c:1698
#12 0x0000000000642911 in sc_asn1_decode (ctx=0x2242830, asn1=0x7ffeb7bd7610, in=0x2242720 "\240!0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", len=262, newp=0x7ffeb7bd7d98, len_left=0x7ffeb7bd7d90) at asn1.c:1749
#13 0x00000000004da9eb in sc_pkcs15_decode_pukdf_entry (p15card=0x22430a0, obj=0x22434e0, buf=0x7ffeb7bd7d98, buflen=0x7ffeb7bd7d90) at pkcs15-pubkey.c:313
#14 0x00000000004b4fe3 in LLVMFuzzerTestOneInput (Data=0x2242720 "\240!0\004V\001\233\a0\r\004\003\023/\001\003\003\001)\004", Size=262) at fuzz_pkcs15_decode.c:57
#15 0x0000000000445672 in ExecuteCallback () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:556
#16 0x0000000000430292 in RunOneTest () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:292
#17 0x0000000000435f4f in FuzzerDriver () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:774
#18 0x00000000004600c3 in main () at /src/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:19
```
Also the `gdb` confirms we are currently handling the `direct` field of `publicECKeyAttributes`:
```
(gdb) f 1
#1  0x00000000006451a3 in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd79e8, obj=0x224273d "\244!\363", objlen=0, depth=5) at asn1.c:1559
(gdb) p *entry
$1 = {name = 0x6718b2 "direct", type = 4, tag = 553648128, flags = 6, parm = 0x2243fa0, arg = 0x2243fa8}
(gdb) f 4
#4  0x0000000000644e4b in asn1_decode_entry (ctx=0x2242830, entry=0x7ffeb7bd78f0, obj=0x224273b "\240", objlen=5, depth=3) at asn1.c:1490
1490				r = asn1_decode(ctx, (struct sc_asn1_entry *) parm, obj,
(gdb) p *entry
$4 = {name = 0x6746c5 "publicECKeyAttributes", type = 129, tag = 16777232, flags = 0, parm = 0x7ffeb7bd7940, arg = 0x0}
```
Actually, my patch would really not address the issue, since as I step further, the decode worked, but later, I see the following:
```
(gdb) p *der
$10 = {value = 0x2e7bcb0 "x\373\017\b@\177", len = 0}
```
So this looks like it is bigger issue that the length of allocation is not propagated to the calling objects. I will try to have a better look into that.

Reply:
The important part is this one:
```
#0  __GI___libc_malloc (bytes=0) at malloc.c:2902
```
The asn1.c calls malloc of size 0, which is invalid, but not reported as error in the fuzzers. This is also the reason why it is not freed in cleanups.

Reply:
@frankmorgner The fixes for this group of problems are already in this branch. I think also for one of the reports from today, which was hitting similar code path.

Reply:
malloc(0) is valid and returns either NULL or a valid pointer, i.e. though it's implementation dependent, you can always free the result... I'm not sure why malloc(0) would cause a leaking memory.

Reply:
Yes, it is valid and it returns either of these two. But in case it returns NULL, the following condition would wrongly return out of memory, usually making all program fail, while zero-length bit string is perfectly valid, I think. It is causing memory leak probably only in the fuzzer implementation. But anyway, we should avoid implementation dependent behavior.

Reply:
I see, thanks for the explanation.